src/main/scala/devices/tilelink/Plic.scala

```diff
@@ -135,71 +136,82 @@ class TLPLIC(params: PLICParams, beatBytes: Int)(implicit p: Parameters) extends
+      else Wire(init=Vec.fill(nDevices)(UInt(1)))
     val threshold =
-      if (nPriorities > 0) Reg(Vec(nHarts, UInt(width = log2Up(nPriorities+1))))
+      if (nPriorities > 0) Reg(Vec(nHarts, UInt(width=prioBits)))
       else Wire(init=Vec.fill(nHarts)(UInt(0)))
-    val pending = Reg(init=Vec.fill(nDevices+1){Bool(false)})
-    val enables = Reg(Vec(nHarts, Vec(nDevices+1, Bool())))
+    val pending = Reg(init=Vec.fill(nDevices){Bool(false)})
+
+    /* Construct the enable registers, chunked into 8-bit segments to reduce verilog size */
+    val firstEnable = nDevices min 7
+    val fullEnables = (nDevices - firstEnable) / 8
+    val tailEnable  = nDevices - firstEnable - 8*fullEnables
+    def enableRegs = (Reg(UInt(width = firstEnable)) +:
+                      Seq.fill(fullEnables) { Reg(UInt(width = 8)) }) ++
+                     (if (tailEnable > 0) Some(Reg(UInt(width = tailEnable))) else None)
+    val enables = Seq.fill(nHarts) { enableRegs }
+    val enableVec = Vec(enables.map(x => Cat(x.reverse)))
+    val enableVec0 = Vec(enableVec.map(x => Cat(x, UInt(0, width=1))))
     
-    def findMax(x: Seq[UInt]): (UInt, UInt) = {
-      if (x.length > 1) {
-        val half = 1 << (log2Ceil(x.length) - 1)
-        val left = findMax(x take half)
-        val right = findMax(x drop half)
-        MuxT(left._1 >= right._1, left, (right._1, UInt(half) | right._2))
-      } else (x.head, UInt(0))
-    }
-
-    val maxDevs = Reg(Vec(nHarts, UInt(width = log2Up(pending.size))))
+    val maxDevs = Reg(Vec(nHarts, UInt(width = log2Ceil(nDevices+1))))
+    val pendingUInt = Cat(pending.reverse :+ UInt(0, width=1))
     for (hart <- 0 until nHarts) {
-      val effectivePriority = (UInt(1) << priority(0).getWidth) +:
-        (for (((p, en), pri) <- (pending zip enables(hart) zip priority).tail)
-          yield Cat(p && en, pri))
-      val (maxPri, maxDev) = findMax(effectivePriority)
-
-      maxDevs(hart) := maxDev
-      harts(hart) := ShiftRegister(Reg(next = maxPri) > Cat(UInt(1), threshold(hart)), params.intStages)
+      val fanin = Module(new PLICFanIn(nDevices, prioBits))
+      fanin.io.prio := priority
+      fanin.io.ip   := enableVec(hart) & pendingUInt
+      maxDevs(hart) := fanin.io.dev
+      harts(hart)   := ShiftRegister(Reg(next = fanin.io.max) > Cat(UInt(1), threshold(hart)), params.intStages)
     }
 
-    def priorityRegDesc(i: Int) = if (i > 0) {
-      RegFieldDesc(s"priority_$i", s"Acting priority of interrupt source $i",
-        reset=if (nPriorities > 0) None else Some(1),
-        wrType=Some(RegFieldWrType.MODIFY))
-    } else {
-      RegFieldDesc.reserved
+    def priorityRegDesc(i: Int) =
+      RegFieldDesc(
+        name      = s"priority_$i",
+        desc      = s"Acting priority of interrupt source $i",
+        group     = Some("priority"),
+        groupDesc = Some("Acting priorities of each interrupt source."),
+        reset     = if (nPriorities > 0) None else Some(1))
```

Comment:
not sure if this is handled below, but to be clear you could say accessType = R or RW accordingly